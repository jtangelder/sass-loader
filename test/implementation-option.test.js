/**
 * @jest-environment node
 */

import nodeSass from 'node-sass';
import dartSass from 'sass';

import {
  compile,
  getTestId,
  getCodeFromBundle,
  getImplementationByName,
  normalizeError,
} from './helpers';

const implementations = [nodeSass, dartSass];

describe('implementation option', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  implementations.forEach((implementation) => {
    const [implementationName] = implementation.info.split('\t');

    it(`${implementationName}`, async () => {
      const nodeSassSpy = jest.spyOn(nodeSass, 'render');
      const dartSassSpy = jest.spyOn(dartSass, 'render');

      const testId = getTestId('language', 'scss');
      const options = {
        implementation: getImplementationByName(implementationName),
      };
      const stats = await compile(testId, { loader: { options } });
      const { css, sourceMap } = getCodeFromBundle(stats);

      expect(css).toBeDefined();
      expect(sourceMap).toBeUndefined();

      expect(stats.compilation.warnings).toMatchSnapshot('warnings');
      expect(stats.compilation.errors).toMatchSnapshot('errors');

      if (implementationName === 'node-sass') {
        expect(nodeSassSpy).toHaveBeenCalledTimes(1);
        expect(dartSassSpy).toHaveBeenCalledTimes(0);
      } else if (implementationName === 'dart-sass') {
        expect(nodeSassSpy).toHaveBeenCalledTimes(0);
        expect(dartSassSpy).toHaveBeenCalledTimes(1);
      }
    });
  });

  it('not specify', async () => {
    const nodeSassSpy = jest.spyOn(nodeSass, 'render');
    const dartSassSpy = jest.spyOn(dartSass, 'render');

    const testId = getTestId('language', 'scss');
    const options = {};
    const stats = await compile(testId, { loader: { options } });
    const { css, sourceMap } = getCodeFromBundle(stats);

    expect(css).toBeDefined();
    expect(sourceMap).toBeUndefined();

    expect(stats.compilation.warnings).toMatchSnapshot('warnings');
    expect(stats.compilation.errors).toMatchSnapshot('errors');

    expect(nodeSassSpy).toHaveBeenCalledTimes(1);
    expect(dartSassSpy).toHaveBeenCalledTimes(0);
  });

  it('should throw an error when the "node-sass" package is an incompatible version', async () => {
    const testId = getTestId('language', 'scss');
    const options = {
      implementation: Object.assign({}, nodeSass, {
        info: 'node-sass\t3.0.0',
      }),
    };

    try {
      const stats = await compile(testId, { loader: { options } });

      getCodeFromBundle(stats);
    } catch (error) {
      expect(normalizeError(error)).toMatchSnapshot();
    }
  });

  it('should throw an error when the "sass" package is an incompatible version', async () => {
    const testId = getTestId('language', 'scss');
    const options = {
      implementation: Object.assign({}, dartSass, {
        info: 'dart-sass\t1.2.0',
      }),
    };

    try {
      const stats = await compile(testId, { loader: { options } });

      getCodeFromBundle(stats);
    } catch (error) {
      expect(normalizeError(error)).toMatchSnapshot();
    }
  });

  it('should throw an error on an unknown sass implementation', async () => {
    const testId = getTestId('language', 'scss');
    const options = {
      implementation: Object.assign({}, dartSass, {
        info: 'strange-sass\t1.0.0',
      }),
    };

    try {
      const stats = await compile(testId, { loader: { options } });

      getCodeFromBundle(stats);
    } catch (error) {
      expect(normalizeError(error)).toMatchSnapshot();
    }
  });

  it('should throw an error when the "info" is unparseable', async () => {
    const testId = getTestId('language', 'scss');
    const options = {
      implementation: Object.assign({}, dartSass, { info: 'asdfj' }),
    };

    try {
      const stats = await compile(testId, { loader: { options } });

      getCodeFromBundle(stats);
    } catch (error) {
      expect(normalizeError(error)).toMatchSnapshot();
    }
  });

  it('should throw an error when the "info" is unparseable #2', async () => {
    const testId = getTestId('language', 'scss');
    const options = {
      implementation: Object.assign({}, dartSass, { info: 'node-sass\t1' }),
    };

    try {
      const stats = await compile(testId, { loader: { options } });

      getCodeFromBundle(stats);
    } catch (error) {
      expect(normalizeError(error)).toMatchSnapshot();
    }
  });

  it('should throw error when the "info" does not exist', async () => {
    const testId = getTestId('language', 'scss');
    const options = {
      // eslint-disable-next-line no-undefined
      implementation: Object.assign({}, dartSass, { info: undefined }),
    };

    try {
      const stats = await compile(testId, { loader: { options } });

      getCodeFromBundle(stats);
    } catch (error) {
      expect(normalizeError(error)).toMatchSnapshot();
    }
  });

  it('should not swallow an error when trying to load a sass implementation', async () => {
    jest.doMock('node-sass', () => {
      const error = new Error('Some error');

      error.code = 'MODULE_NOT_FOUND';
      error.stack = null;

      throw error;
    });

    jest.doMock('sass', () => {
      const error = new Error('Some error');

      error.code = 'MODULE_NOT_FOUND';
      error.stack = null;

      throw error;
    });

    const testId = getTestId('language', 'scss');
    const options = {};

    try {
      const stats = await compile(testId, { loader: { options } });

      getCodeFromBundle(stats);
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
});
