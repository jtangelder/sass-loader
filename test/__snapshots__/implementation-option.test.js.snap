// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`implementation option dart-sass: errors 1`] = `Array []`;

exports[`implementation option dart-sass: warnings 1`] = `Array []`;

exports[`implementation option node-sass: errors 1`] = `Array []`;

exports[`implementation option node-sass: warnings 1`] = `Array []`;

exports[`implementation option not specify: errors 1`] = `Array []`;

exports[`implementation option not specify: warnings 1`] = `Array []`;

exports[`implementation option should not swallow an error when trying to load a sass implementation 1`] = `
[Error: Module build failed (from ../src/cjs.js):
Some error]
`;

exports[`implementation option should throw an error on an unknown sass implementation 1`] = `
[Error: Module build failed (from ../src/cjs.js):
Error: Unknown Sass implementation "strange-sass".
    at ReplacedStackEntry
    at ReplacedStackEntry]
`;

exports[`implementation option should throw an error when the "info" is unparseable #2 1`] = `
[Error: Module build failed (from ../src/cjs.js):
Error: Invalid Sass version "1".
    at ReplacedStackEntry
    at ReplacedStackEntry]
`;

exports[`implementation option should throw an error when the "info" is unparseable 1`] = `
[Error: Module build failed (from ../src/cjs.js):
Error: Unknown Sass implementation "asdfj".
    at ReplacedStackEntry
    at ReplacedStackEntry]
`;

exports[`implementation option should throw an error when the "node-sass" package is an incompatible version 1`] = `
[Error: Module build failed (from ../src/cjs.js):
Error: Node Sass version 3.0.0 is incompatible with ^4.0.0.
    at ReplacedStackEntry
    at ReplacedStackEntry]
`;

exports[`implementation option should throw an error when the "sass" package is an incompatible version 1`] = `
[Error: Module build failed (from ../src/cjs.js):
Error: Dart Sass version 1.2.0 is incompatible with ^1.3.0.
    at ReplacedStackEntry
    at ReplacedStackEntry]
`;

exports[`implementation option should throw error when the "info" does not exist 1`] = `
[Error: Module build failed (from ../src/cjs.js):
Error: Unknown Sass implementation.
    at ReplacedStackEntry
    at ReplacedStackEntry]
`;
